solr/core/src/java/org/apache/solr/packagemanager/PackageUtils.java

```diff
@@ -94,29 +92,34 @@ public static void postFile(SolrClient client, ByteBuffer buffer, String name, S
     if (sig != null) {
       params.add("sig", sig);
     }
-    V2Response rsp =
-        new V2Request.Builder(resource)
-            .withMethod(SolrRequest.METHOD.PUT)
-            .withPayload(buffer)
-            .forceV2(true)
-            .withMimeType("application/octet-stream")
-            .withParams(params)
-            .build()
-            .process(client);
-    if (!name.equals(rsp.getResponse().get(CommonParams.FILE))) {
-      throw new SolrException(
-          ErrorCode.BAD_REQUEST,
-          "Mismatch in file uploaded. Uploaded: "
-              + rsp.getResponse().get(CommonParams.FILE)
-              + ", Original: "
-              + name);
-    }
+    GenericSolrRequest request =
+        new GenericSolrRequest(SolrRequest.METHOD.PUT, resource, params) {
+          @Override
+          public RequestWriter.ContentWriter getContentWriter(String expectedType) {
+            return new RequestWriter.ContentWriter() {
+              public final ByteBuffer payload = buffer;
+
+              @Override
+              public void write(OutputStream os) throws IOException {
+                if (payload == null) return;
+                os.write(payload.array());
```

Comment:
array() is an optional method, and even if it it implemented, you'd still need to respect the offset & length.
Instead: https://stackoverflow.com/a/579616/92186